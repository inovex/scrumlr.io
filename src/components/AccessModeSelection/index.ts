import AccessModeSelection from './AccessModeSelection';
export default AccessModeSelection;
export * from './AccessModeSelection';
