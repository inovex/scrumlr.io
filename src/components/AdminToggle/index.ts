export * from './AdminToggle';

import AdminToggle from './AdminToggle';
export default AdminToggle;
