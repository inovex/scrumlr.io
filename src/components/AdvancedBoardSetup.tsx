import React from 'react';
import Typography from '@material-ui/core/Typography';
import FormGroup from '@material-ui/core/FormGroup';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Switch from '@material-ui/core/Switch';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import { useTranslation } from 'react-i18next';

export interface AdvancedBoardSetupProps {
  admissionControl: boolean;
  encryptedData: boolean;
  onSetAdmissionControl: (enabled: boolean) => void;
  onSetEncryptedData: (enabled: boolean) => void;
}

export const AdvancedBoardSetup: React.FC<AdvancedBoardSetupProps> = ({ admissionControl, encryptedData, onSetAdmissionControl, onSetEncryptedData }) => {
  const { t } = useTranslation('advancedBoardSetup');

  return (
    <>
      <ExpansionPanel>
        <ExpansionPanelSummary>
          <Typography>{t('heading')}</Typography>
        </ExpansionPanelSummary>
        <ExpansionPanelDetails>
          <div>
            <Typography>{t('helper')}</Typography>
            <FormGroup>
              <FormControlLabel
                onClick={(event) => event.stopPropagation()}
                onFocus={(event) => event.stopPropagation()}
                control={
                  <Switch
                    checked={admissionControl}
                    onChange={(event) => {
                      onSetAdmissionControl(event.target.checked);
                    }}
                  />
                }
                label={t('admissionControl.heading')}
              />
              <Typography color="textSecondary">{t('admissionControl.helper')}</Typography>

              <FormControlLabel
                onClick={(event) => event.stopPropagation()}
                onFocus={(event) => event.stopPropagation()}
                control={<Switch checked={encryptedData} onChange={(event) => onSetEncryptedData(event.target.checked)} disabled={!admissionControl} />}
                label={t('encryptData.heading')}
              />
              <Typography color="textSecondary">{t('encryptData.helper')}</Typography>
            </FormGroup>
          </div>
        </ExpansionPanelDetails>
      </ExpansionPanel>
    </>
  );
};

export default AdvancedBoardSetup;
