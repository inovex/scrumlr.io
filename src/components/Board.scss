@import "common";

.board {
  --n: 1;

  display:grid;
  grid-auto-flow: column;
  grid-auto-columns: calc(100%/min(var(--real-columns), var(--n)));
  overflow-x:auto;
  scroll-snap-type: x mandatory;

  height: 100vh;
}

@for $i from 1 through 12 {
  @media screen and (min-width: $i*$min-column-width) {
    .board {
      --n: #{$i};
    }
  }
}