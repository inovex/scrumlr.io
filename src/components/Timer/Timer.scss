@import "constants/style.scss";

$timer-display-width: 114px;
$timer__content-height: 38px;
$timer__short-actions-height: 28px;

$timer__progress-bar-color: #6096ff;

.timer__container {
  height: 100%;
  display: flex;
  flex-direction: row;
  gap: 8px;
}

.timer__increment-button {
  position: relative;
  border-radius: 999px;

  &:focus-within {
    outline: 2px solid $color-backlog-blue;
  }
}

.increment-button__button {
  border: 2px solid $color-white;
  background-color: $color-dark-mode-note;
  box-shadow: $box-shadow--light;
  border-radius: 999px;
  height: 42px;
  width: 42px;
  color: white;
  cursor: pointer;
  display: grid;
  place-content: center;
  outline: none;

  > svg {
    width: 100%;
  }
}

.timer {
  height: 100%;
  width: $timer-display-width;
  color: $color-dark-mode-note;
  background-color: $menu-icon-background-color--light;
  border-radius: $rounded--full;
  position: relative;
  border: 2px solid $menu-icon-background-color--light;
  box-shadow: $box-shadow--light;
}
.timer > svg {
  position: absolute;
  height: 100%;
  background-color: $color-dark-mode-note;
  color: $color-white;
  border-radius: 100%;
}
[theme="dark"] .timer > svg {
  background-color: $color-white;
  color: $color-dark-mode-note;
}

.timer > span {
  position: absolute;
  right: 0;
  left: 0;
  height: 100%;
  width: 100%;
  padding-left: $timer__content-height;
  line-height: $timer__content-height;
  text-align: center;
  font-weight: bold;
}

.timer__progress-bar {
  position: absolute;
  top: 0;
  left: calc($timer__content-height / 2);
  height: $timer__content-height;
  border-top-right-radius: $rounded--full;
  border-bottom-right-radius: $rounded--full;
  background-color: $timer__progress-bar-color;
  transition: right 250ms linear;
}

.timer__short-actions {
  height: 100%;
  width: 100%;
  position: absolute;
  border-radius: $rounded--full;
  background-color: $menu-icon-background-color--light;
  display: none;
  justify-content: end;
  align-items: center;
  gap: $margin--small;
  padding-right: 4px;
}
[theme="dark"] .timer__short-actions {
  background-color: $menu-icon-background-color--dark;
}

.timer:hover > .timer__short-actions {
  display: flex;
}

.short-actions__short-action {
  position: relative;
  height: $timer__short-actions-height;
  width: $timer__short-actions-height;
  border-radius: $rounded--full;
}

.short-action__button {
  height: $timer__short-actions-height;
  width: $timer__short-actions-height;
  border-radius: $rounded--full;
  background-color: $color-lighter-gray;
  color: $color-dark-mode-note;
  border: none;
  padding: 0;
  cursor: pointer;
  display: flex;
  justify-content: center;
  align-items: center;
}
[theme="dark"] .short-action__button {
  background-color: $color-dark-one;
  color: $color-white;
}
.short-action__button > svg {
  position: absolute;
  height: $icon--small;
  width: $icon--small;
}
.short-action__button--ready > .short-action__cancel-icon,
.short-action__button--unready > .short-action__cancel-icon {
  visibility: hidden;
}
.short-action__button--ready:hover > .short-action__check-icon {
  visibility: hidden;
}
.short-action__button--ready:hover > .short-action__cancel-icon {
  visibility: visible;
}
.short-action__button > .short-action__check-icon {
  height: $timer__short-actions-height;
  width: $timer__short-actions-height;
}

.short-action__button:hover {
  background-color: $timer__progress-bar-color;
  color: $color-white;
}

.short-action__tooltip,
.increment-button__tooltip {
  width: max-content;
  visibility: hidden;
  opacity: 0;
  transition: opacity 150ms linear;
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
  border-radius: $rounded--full;
  padding: $padding--small $padding--default;
  background-color: $tooltip-background-color--light;
  color: $menu-icon-background-color--dark;
  font-size: $text--base;
  font-weight: bold;
  bottom: calc(100% + 6px);
  box-shadow:
    0 1px 3px 0 rgb(0 0 0 / 0.1),
    0 1px 2px -1px rgb(0 0 0 / 0.1);
  // Increased the z-index to make sure the tooltip doesn't get overlapped by the absolute positioned icon
  z-index: calc($infobar-z-index + 50);
}
@media (prefers-reduced-motion) {
  .short-action__tooltip {
    transition: none;
  }
}

.timer__increment-button:hover > .increment-button__tooltip,
.short-actions__short-action:hover > .short-action__tooltip {
  visibility: visible;
  opacity: 1;
}

.increment-button__tooltip::after,
.short-action__tooltip::after {
  content: "";
  position: absolute;
  border: 5px solid;
  border-color: $tooltip-background-color--light transparent transparent transparent;
  left: 50%;
  transform: translateX(-50%);
  top: 100%;
}

[theme="dark"] {
  .timer {
    color: $color-white;
    background-color: $menu-icon-background-color--dark;
    border-color: $menu-icon-background-color--dark;
  }

  .timer__increment-button:focus-within {
    outline: 2px solid $color-planning-pink;
  }

  .increment-button__button {
    background-color: $color-white;
    color: $color-dark-mode-note;
    border-color: $menu-icon-background-color--dark;
  }

  .short-action__tooltip,
  .increment-button__tooltip {
    background-color: $tooltip-background-color--dark;
    color: $color-white;

    &::after {
      border-color: $tooltip-background-color--dark transparent transparent transparent;
    }
  }
}

// Shake animation on expired timer
.timer--expired {
  animation: shake 0.5s;
  animation-iteration-count: 10;
}
@media (prefers-reduced-motion: reduce) {
  .timer--expired > span {
    animation: none;
  }
}
@keyframes shake {
  0% {
    transform: translate(1px, 1px) rotate(0deg);
  }
  10% {
    transform: translate(-1px, -2px) rotate(-1deg);
  }
  20% {
    transform: translate(-3px, 0px) rotate(1deg);
  }
  30% {
    transform: translate(3px, 2px) rotate(0deg);
  }
  40% {
    transform: translate(1px, -1px) rotate(1deg);
  }
  50% {
    transform: translate(-1px, 2px) rotate(-1deg);
  }
  60% {
    transform: translate(-3px, 1px) rotate(0deg);
  }
  70% {
    transform: translate(3px, 1px) rotate(-1deg);
  }
  80% {
    transform: translate(-1px, -1px) rotate(1deg);
  }
  90% {
    transform: translate(1px, 2px) rotate(0deg);
  }
  100% {
    transform: translate(1px, -2px) rotate(-1deg);
  }
}
