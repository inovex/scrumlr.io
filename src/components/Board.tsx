import React from 'react';
import './Board.scss';
import {getColorClassName} from "../constants/colors";

const Board = ({ children }: any) => {
    const colors = React.Children.map(children, (child) => child.props.color);

    return (
        <>
            <style>
                {`.board { --board__columns: ${React.Children.count(children)} }`}
            </style>
            <div className="board">
                <div className={`board__spacer-left ${getColorClassName(colors[0])}`} />
                    {children}
                <div className={`board__spacer-right ${getColorClassName(colors[colors.length - 1])}`} />
            </div>
        </>
    )
};

export default Board;