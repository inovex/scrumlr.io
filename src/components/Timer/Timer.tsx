import {useEffect, useRef, useState} from "react";
import classNames from "classnames";
import store, {useAppSelector} from "store";
import {Actions} from "store/action";
import {ReactComponent as CancelIcon} from "assets/icon-cancel.svg";
import {ReactComponent as TimerIcon} from "assets/icon-timer.svg";
import {ReactComponent as CheckIcon} from "assets/icon-check.svg";
import {ReactComponent as PlusOneIcon} from "assets/icon-plus-one.svg";
import {ReactComponent as FlagIcon} from "assets/icon-flag.svg";
import {useTranslation} from "react-i18next";
import {Toast} from "utils/Toast";
import useSound from "use-sound";
import {API} from "api";
import {Timer as TimerUtils} from "utils/timer";
import "./Timer.scss";
import {TOAST_TIMER_DEFAULT} from "constants/misc";

type TimerProps = {
  startTime: Date;
  endTime: Date;
};

const usePrevious = (value: boolean) => {
  const ref = useRef<boolean>();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

export const Timer = (props: TimerProps) => {
  const {t} = useTranslation();

  const allParticipantsReady = useAppSelector(
    (state) =>
      state.participants!.others.filter((p) => p.connected && p.role === "PARTICIPANT").length &&
      state.participants!.others.filter((p) => p.connected && p.role === "PARTICIPANT").every((participant) => participant.ready)
  );
  const anyReady = useAppSelector((state) => state.participants!.others.filter((p) => p.connected).some((participant) => participant.ready));
  const isModerator = useAppSelector((state) => state.participants?.self.role === "OWNER" || state.participants?.self.role === "MODERATOR");
  const isReady = useAppSelector((state) => state.participants?.self.ready);
  const boardId = useAppSelector((state) => state.board.data!.id);
  const me = useAppSelector((state) => state.participants?.self);

  const [playTimesUpSound, {sound: timesUpSoundObject}] = useSound(`${process.env.PUBLIC_URL}/timer_finished.mp3`, {volume: 0.5, interrupt: true});
  const [timeLeft, setTimeLeft] = useState<{h: number; m: number; s: number}>(TimerUtils.calculateTimeLeft(props.endTime));
  const [elapsedTimePercentage, setElapsedTimePercentage] = useState<number>(TimerUtils.calculateElapsedTimePercentage(props.startTime, props.endTime));
  const [timesUpShouldPlay, setTimesUpShouldPlay] = useState(false);
  const [playTimesUp, setPlayTimesUp] = useState(false);
  const previousPlayTimesUpState = usePrevious(playTimesUp);

  useEffect(() => {
    const timerUpdateTimeout = setTimeout(() => {
      setTimeLeft(TimerUtils.calculateTimeLeft(props.endTime));
      setElapsedTimePercentage(TimerUtils.calculateElapsedTimePercentage(props.startTime, props.endTime));
    }, 250);
    return () => clearTimeout(timerUpdateTimeout);
  });

  useEffect(() => {
    if (!previousPlayTimesUpState && playTimesUp) {
      timesUpSoundObject.on("end", () => setPlayTimesUp(false));
      playTimesUpSound();
      if (!isModerator && !isReady) {
        Toast.info({
          title: t("Toast.ready"),
          buttons: [t("Toast.readyButton")],
          firstButtonOnClick: () => store.dispatch(Actions.setUserReadyStatus(me!.user.id, true)),
        });
      }
      if (isModerator && anyReady) {
        Toast.info({
          title: t("Toast.moderatorResetReadyStates"),
          buttons: [t("Toast.moderatorResetReadyStatesButton")],
          firstButtonOnClick: () => API.updateReadyStates(boardId, false),
          autoClose: false,
        });
      }
    }

    return () => {};
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [playTimesUp]);

  useEffect(() => {
    if (timeLeft.m === 0 && !playTimesUp) {
      if (timeLeft.s <= 0) {
        if (timesUpShouldPlay) {
          setPlayTimesUp(true);
          setTimesUpShouldPlay(false);
        }
      } else if (!timesUpShouldPlay) {
        setTimesUpShouldPlay(true);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [timeLeft]);

  useEffect(() => {
    if (isModerator && allParticipantsReady && Object.values(timeLeft).some((time) => time > 0)) {
      Toast.info({title: t("Toast.allParticipantsDone"), autoClose: TOAST_TIMER_DEFAULT});
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allParticipantsReady, isModerator]);

  return (
    <div className="timer__container">
      <div id="timer" className={classNames("timer", {"timer--expired": timeLeft.m === 0 && timeLeft.s === 0})}>
        <div className="timer__progress-bar" style={{right: `calc(72px - ${elapsedTimePercentage} * 72px)`}} />
        <span>
          {String(timeLeft!.m).padStart(2, "0")}:{String(timeLeft!.s).padStart(2, "0")}
        </span>
        <div className="timer__short-actions">
          {isModerator && (
            <div className="short-actions__short-action">
              <button aria-label={t("Timer.endTimer")} className="short-action__button" onClick={() => store.dispatch(Actions.cancelTimer())}>
                <FlagIcon />
              </button>
              <span className="short-action__tooltip">{t("Timer.endTimer")}</span>
            </div>
          )}
          <div className="short-actions__short-action">
            <button
              aria-label={isReady ? t("MenuBars.unmarkAsDone") : t("MenuBars.markAsDone")}
              className={classNames("short-action__button", {"short-action__button--ready": isReady}, {"short-action__button--unready": !isReady})}
              onClick={() => store.dispatch(Actions.setUserReadyStatus(me!.user.id, !isReady))}
            >
              <CheckIcon className="short-action__check-icon" />
              <CancelIcon className="short-action__cancel-icon" />
            </button>
            <span className="short-action__tooltip">{isReady ? t("MenuBars.unmarkAsDone") : t("MenuBars.markAsDone")}</span>
          </div>
        </div>
        <TimerIcon />
      </div>
      {isModerator && (
        <div className="timer__increment-button">
          <button aria-label={t("Timer.addOneMinute")} className="increment-button__button" onClick={() => store.dispatch(Actions.incrementTimer())}>
            <PlusOneIcon />
          </button>
          <span className="increment-button__tooltip">{t("Timer.addOneMinute")}</span>
        </div>
      )}
    </div>
  );
};
