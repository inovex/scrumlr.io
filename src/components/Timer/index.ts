export * from './Timer';

import Timer from './Timer';
export default Timer;
