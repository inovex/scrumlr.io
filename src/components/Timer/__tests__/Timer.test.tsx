import {render} from "testUtils";
import {Timer} from "components/Timer";
import {useAppSelector} from "store";
import {t} from "i18next";

jest.mock("store");
const mockedUseAppSelector = jest.mocked(useAppSelector);

jest.useFakeTimers().setSystemTime(new Date(60000));

describe("Timer", () => {
  test("should render correctly", () => {
    mockedUseAppSelector.mockResolvedValue({} as never);
    const {container} = render(<Timer startTime={new Date(0)} endTime={new Date(180000)} />, {container: document.getElementById("root")!});
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should have cancel button if user is moderator", () => {
    mockedUseAppSelector.mockResolvedValue({} as never);
    const {container} = render(<Timer startTime={new Date(0)} endTime={new Date(180000)} />, {container: document.getElementById("root")!});
    expect(container.firstChild).toContainElement(container.getElementsByTagName("button")[0]);
  });

  test("should have expired class if timer is over", () => {
    mockedUseAppSelector.mockResolvedValue({} as never);
    const {container} = render(<Timer startTime={new Date(0)} endTime={new Date(60000)} />, {container: document.getElementById("root")!});
    expect(container.getElementsByClassName("timer--expired").length).toBe(1);
  });

  test("should have increment button if user is moderator", () => {
    mockedUseAppSelector.mockResolvedValue({} as never);
    const {getByText} = render(<Timer startTime={new Date(0)} endTime={new Date(60000)} />, {container: document.getElementById("root")!});
    expect(getByText(t("Timer.addOneMinute"))).toBeDefined();
  });
});
