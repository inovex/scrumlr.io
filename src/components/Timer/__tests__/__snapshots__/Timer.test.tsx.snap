// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should render correctly 1`] = `
<div
  class="timer__container"
>
  <div
    class="timer"
    id="timer"
  >
    <div
      class="timer__progress-bar"
      style="right: calc(72px - 0.6666666666666667 * 72px);"
    />
    <span>
      02
      :
      00
    </span>
    <div
      class="timer__short-actions"
    >
      <div
        class="short-actions__short-action"
      >
        <button
          aria-label="End timer"
          class="short-action__button"
        >
          <svg>
            icon-flag.svg
          </svg>
        </button>
        <span
          class="short-action__tooltip"
        >
          End timer
        </span>
      </div>
      <div
        class="short-actions__short-action"
      >
        <button
          aria-label="Unmark me as done"
          class="short-action__button short-action__button--ready"
        >
          <svg
            class="short-action__check-icon"
          >
            icon-check.svg
          </svg>
          <svg
            class="short-action__cancel-icon"
          >
            icon-cancel.svg
          </svg>
        </button>
        <span
          class="short-action__tooltip"
        >
          Unmark me as done
        </span>
      </div>
    </div>
    <svg>
      icon-timer.svg
    </svg>
  </div>
  <div
    class="timer__increment-button"
  >
    <button
      aria-label="Increase timer by one minute"
      class="increment-button__button"
    >
      <svg>
        icon-plus-one.svg
      </svg>
    </button>
    <span
      class="increment-button__tooltip"
    >
      Increase timer by one minute
    </span>
  </div>
</div>
`;
